File: day11/main.go
package main

import (
	"fmt"
)

const (
	puzzleInput = 9810

	boardSize = 300
)

func powerLevel(x, y int) int {
	rackID := x + 10
	return (rackID*y+puzzleInput)*rackID/100%10 - 5
}

func generatePower() [][]int {
	board := make([][]int, boardSize)
	for i := range board {
		board[i] = make([]int, boardSize)
	}
	for y, i := range board {
		for x, _ := range i {
			board[y][x] = powerLevel(x+1, y+1)
		}
	}
	return board
}

func generatePartials(board [][]int) [][]int {
	partial := make([][]int, boardSize+1)
	for i := range partial {
		partial[i] = make([]int, boardSize+1)
	}
	for y := 1; y < boardSize+1; y++ {
		for x := 1; x < boardSize+1; x++ {
			partial[y][x] = board[y-1][x-1] + partial[y][x-1] + partial[y-1][x] - partial[y-1][x-1]
		}
	}
	return partial
}

func main() {
	board := generatePower()
	partial := generatePartials(board)
	{
		maxPower := 0
		maxX := 0
		maxY := 0
		for y := 0; y < boardSize-3; y++ {
			for x := 0; x < boardSize-3; x++ {
				power := partial[y+3][x+3] - partial[y][x+3] - partial[y+3][x] + partial[y][x]
				if power > maxPower {
					maxPower = power
					maxX = x + 1
					maxY = y + 1
				}
			}
		}
		fmt.Println(maxPower, maxX, maxY)
	}

	{
		maxPower := 0
		maxX := 0
		maxY := 0
		maxSize := 0
		for i := 1; i < boardSize+1; i++ {
			for y := 0; y < boardSize-i+1; y++ {
				for x := 0; x < boardSize-i+1; x++ {
					power := partial[y+i][x+i] - partial[y][x+i] - partial[y+i][x] + partial[y][x]
					if power > maxPower {
						maxPower = power
						maxX = x + 1
						maxY = y + 1
						maxSize = i
					}
				}
			}
		}
		fmt.Println(maxPower, maxX, maxY, maxSize)
	}
}

File: day11/src/main.rs
const PUZZLE_INPUT: isize = 9810;
const BOARD_SIZE: usize = 300;

fn power_level(x: isize, y: isize) -> isize {
    let id = x + 10;
    (id * y + PUZZLE_INPUT) * id / 100 % 10 - 5
}

fn main() {
    let board = {
        let mut board = vec![vec![0; BOARD_SIZE]; BOARD_SIZE];
        for (y, row) in board.iter_mut().enumerate() {
            for (x, i) in row.iter_mut().enumerate() {
                *i = power_level((x + 1) as isize, (y + 1) as isize)
            }
        }
        board
    };
    let partial = {
        let mut partial = vec![vec![0; BOARD_SIZE + 1]; BOARD_SIZE + 1];
        for y in 1..partial.len() {
            for x in 1..partial.len() {
                partial[y][x] = board[y - 1][x - 1] + partial[y][x - 1] + partial[y - 1][x]
                    - partial[y - 1][x - 1];
            }
        }
        partial
    };
    {
        let mut mp = 0;
        let mut mx = 0;
        let mut my = 0;
        for y in 0..partial.len() - 3 {
            for x in 0..partial.len() - 3 {
                let power =
                    partial[y + 3][x + 3] - partial[y][x + 3] - partial[y + 3][x] + partial[y][x];
                if power > mp {
                    mp = power;
                    mx = x + 1;
                    my = y + 1;
                }
            }
        }
        println!("{} {} {}", mp, mx, my);
    }
    {
        let mut mp = 0;
        let mut mx = 0;
        let mut my = 0;
        let mut ms = 0;
        for i in 1..partial.len() {
            for y in 0..partial.len() - i {
                for x in 0..partial.len() - i {
                    let power = partial[y + i][x + i] - partial[y][x + i] - partial[y + i][x]
                        + partial[y][x];
                    if power > mp {
                        mp = power;
                        mx = x + 1;
                        my = y + 1;
                        ms = i;
                    }
                }
            }
        }
        println!("{} {} {} {}", mp, mx, my, ms);
    }
}
